### PR TITLE
retrive detail error info for connection failure by set 'errno'

### PR DESCRIPTION
**Pull request by chenbd**
so that on_connect() callback can see what happened by read 'errno'

### CODE REVIEW

File: examples/libh2o/httpclient.c [hunk 1]
```diff
@@ -229,7 +229,9 @@ h2o_httpclient_head_cb on_connect(h2o_httpclient_t *client, const char *errstr,
                                   h2o_url_t *origin)
 {
     if (errstr != NULL) {
-        on_error(client->ctx, errstr);
+        char buf[128];
+        on_error(client->ctx, "%s:%s", errstr,
+                 h2o_strerror_r(errno, buf, sizeof(buf)));
```
Comment (kazuho) [inline examples/libh2o/httpclient.c]: I do not think that we should be using `h2o_strerror_r` in the HTTP-level callbacks. There is no guarantee that the cause of `on_connect` failing is due to an error returned from a syscall (consider TLS certificate mismatch for an example).

File: lib/common/socketpool.c [hunk 2]
```diff
@@ -376,7 +376,7 @@ static void on_connect(h2o_socket_t *sock, const char *err)
         }
         __sync_sub_and_fetch(&req->pool->_shared.count, 1);
         req->sock = NULL;
-        errstr = "connection failed"; /* shouldn't we return err? */
+        errstr = err;
```
Comment (kazuho) [inline lib/common/socketpool.c]: This looks like a fine change, though I think we can just get rid of `errstr`, and pass `err` as the argument to `call_connect_cb`?

File: lib/common/socketpool.c [hunk 3]
```diff
@@ -410,7 +410,7 @@ static void start_connect(h2o_socketpool_connect_request_t *req, struct sockaddr
             return;
         }
         __sync_sub_and_fetch(&req->pool->_shared.count, 1);
-        call_connect_cb(req, "failed to connect to host");
+        call_connect_cb(req, h2o_socket_error_conn_fail);
```
Comment (kazuho) [inline lib/common/socketpool.c]: This looks OK too.

File: examples/libh2o/httpclient.c (hunk 1 above)
Comment (chenbd) [reply examples/libh2o/httpclient.c]: Yes, `on_connect` callback will be delayed after ssl handshake finished(or error) when using ssl for http layer.
How about do this in socketpool's on `on_connect`?

Comment (kazuho) [reply examples/libh2o/httpclient.c]: That would collide with the current design, that uses the address of `errstr` to see what the error was.

The socket layer is expected define it's own error strings, and declare them as `const char *h2o_socket_error_XXX`. Applications are expected to compare the address of the returned string if it needs to take action based on the specific error that happened.

Comment (chenbd) [reply examples/libh2o/httpclient.c]: Thanks for your clarification.
Seems currently applications just do the `NULL` judge and do not compare with socket layer error strings.
Anyway, we can leave it as it is.

File: examples/libh2o/latency-optimization.c [hunk 4]
```diff
@@ -202,8 +202,10 @@ static void on_handshake_complete(h2o_socket_t *sock, const char *err)
 static void on_connect(h2o_socket_t *sock, const char *err)
 {
     if (err != NULL) {
+        char buf[128];
         /* connection failed */
-        fprintf(stderr, "failed to connect to host:%s\n", err);
+        fprintf(stderr, "failed to connect to host:%s:%s\n", err,
+                h2o_strerror_r(errno, buf, sizeof(buf)));
```
Comment (kazuho) [inline examples/libh2o/latency-optimization.c]: I think this change is also not appropriate due to the same reason (i.e., the failure might be due to something other than a syscall, e.g., certificate validation).

File: lib/common/socket/evloop.c.h [hunk 5]
```diff
@@ -530,9 +530,11 @@ static void run_socket(struct st_h2o_evloop_socket_t *sock)
             int so_err = 0;
             socklen_t l = sizeof(so_err);
             so_err = 0;
-            if (getsockopt(sock->fd, SOL_SOCKET, SO_ERROR, &so_err, &l) != 0 || so_err != 0) {
-                /* FIXME lookup the error table */
+            if (getsockopt(sock->fd, SOL_SOCKET, SO_ERROR, &so_err, &l) != 0) {
                 err = h2o_socket_error_conn_fail;
+            } else if (so_err != 0) {
+                err = h2o_socket_error_conn_fail;
+                errno = so_err;
```
Comment (kazuho) [inline lib/common/socket/evloop.c.h]: I do not think we need to set `errno`, due to the same reason as above.

If we want to notify a fine grained error from here, it should be done by adding more string constants to `h2o_socket_error_*` and setting one of them in `err`.

Comment (chenbd) [reply lib/common/socket/evloop.c.h]: Guess 'h2o_socket_error_conn_fail' is enough for upper layer or the user.and there are many reasons for connection failure. So errno is just a hint for the reason.just like Read/write error.
If not set here, the reason can be retrieved by the user manually.

File: examples/libh2o/latency-optimization.c (hunk 4 above)
Comment (chenbd) [reply examples/libh2o/latency-optimization.c]: Seems 'on_connect' only called after connection ready before ssl handshake, am i misunderstanding something?

File: lib/common/socket/evloop.c.h (hunk 5 above)
Comment (kazuho) [reply lib/common/socket/evloop.c.h]: Well, the thing is that an application cannot tell if `errno` reflects an error of the socket that it is dealing with. We do not have such a contract (that `error` reflects the state of the socket when some callbacks are called), and we cannot have such a contract especially when libuv is used as a event loop.

Comment (chenbd) [reply lib/common/socket/evloop.c.h]: Thanks for your patient explanation.
`getsockopt(SO_ERROR)` should be used by application when needed to obtain the socket error.
guess this also applied to libuv.

**Comment by kazuho** [comment]
Thank you for the changes. Merged to master.

**Pull request merged**